Bottom: 92efe862e074254f5fdca04862f8b37d5730be9f
Top:    02c5f92e556da6b3a95c0d3523fd89f84202ad1c
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-06-06 20:27:01 +0100

Introduce POLYGONHOLE_MODE for creating holes in polygons

Having selected polygon hole mode, the first click selects which
polygon to cut a hole in. A second click defines the start point of
the hole contour. The tool then behaves in a similar way to the
polygon drawing tool, with the hole ending when the start point is
re-clicked.

To avoid creating illegal polygons, the hole drawn is subtracted from
a representation of the original polygon with the poly_Boolean_free().
This consolidates any contours it intersects with and prevents the user
defining contours which intersect each other. (Although we don't
currently prevent the the user drawing self-intersecting contours).

The resulting POLYAREA is re-processed into PolygonType objects,
potentially more than one - if the hole drawn bisects the original
polygon. To keep undo operations simple, these are added as completely
new objects and the original polygon is deleted - along with its ID.


---

diff --git a/src/gpcb-menu.res b/src/gpcb-menu.res
index dde5c2d..c4cb62a 100644
--- a/src/gpcb-menu.res
+++ b/src/gpcb-menu.res
@@ -517,6 +517,7 @@ PopupMenus =
        {"Copy" checked=copymode,1 Mode(Copy)}
        {"Lock" checked=lockmode,1 Mode(Lock) a={"F12" "<Key>F12"}}
        {"Cancel" Mode(Escape) a={"Esc" "<Key>Escape"}}
+       {"PolygonHole" checked=polygonholemode,1 Mode(PolygonHole)}
       }
     }
   }
diff --git a/src/pcb-menu.res b/src/pcb-menu.res
index 0df4b7f..e188686 100644
--- a/src/pcb-menu.res
+++ b/src/pcb-menu.res
@@ -185,6 +185,7 @@ MainMenu =
    {"Copy" checked=copymode,1 Mode(Copy)}
    {"Lock" checked=lockmode,1 Mode(Lock) a={"F12" "<Key>F12"}}
    {"Cancel" Mode(Cancel) a={"Esc" "<Key>Escape"}}
+   {"PolygonHole" checked=polygonholemode,1 Mode(PolygonHole)}
    -
    {"Command" Command() a={":" "<Key>:"}}
   }
